server/src/main/java/org/opensearch/cluster/routing/FailAwareWeightedRouting.java

```diff
@@ -0,0 +1,155 @@
+
+    /* exception causing failure for previous shard copy */
+    private Exception exception;
+
+    private ClusterState clusterState;
+
+    public FailAwareWeightedRouting(Exception e, ClusterState clusterState) {
+        this.exception = e;
+        this.clusterState = clusterState;
+    }
+
+    /**
+     * *
+     * @return true if exception is due to cluster availability issues
+     */
+    private boolean isInternalFailure() {
+        if (exception instanceof OpenSearchException) {
+            // checking for 5xx failures
+            return (((OpenSearchException) exception).status().getStatus() / 100) * 100 == RestStatus.INTERNAL_SERVER_ERROR.getStatus();
+        }
+        return false;
+    }
+
+    /**
+     * This function checks if the shard is present in data node with weighted routing weight set to 0,
+     * In such cases we fail open, if shard search request for the shard from other shard copies fail with non
+     * retryable exception.
+     *
+     * @param nodeId the node with the shard copy
+     * @return true if the node has attribute value with shard routing weight set to zero, else false
+     */
+    private boolean isWeighedAway(String nodeId) {
+        DiscoveryNode node = clusterState.nodes().get(nodeId);
+        WeightedRoutingMetadata weightedRoutingMetadata = clusterState.metadata().weightedRoutingMetadata();
+        if (weightedRoutingMetadata != null) {
+            WeightedRouting weightedRouting = weightedRoutingMetadata.getWeightedRouting();
+            // TODO: check weighted routing has weights set after merging versioning changes
+            if (weightedRouting != null) {
+                // Fetch weighted routing attributes with weight set as zero
+                Stream<String> keys = weightedRouting.weights()
+                    .entrySet()
+                    .stream()
+                    .filter(entry -> entry.getValue().intValue() == 0)
+                    .map(Map.Entry::getKey);
+
+                for (Object key : keys.toArray()) {
+                    if (node.getAttributes().get(weightedRouting.attributeName()).equals(key.toString())) {
+                        return true;
+                    }
+                }
+            }
+        }
+        return false;
+    }
+
+    /**
+     * This function returns next shard copy to retry search request in case of failure from previous copy returned
+     * by the iterator. It has the logic to fail open ie request shard copies present in nodes with weighted shard
+     * routing weight set to zero
+     *
+     * @param shardIt Shard Iterator containing order in which shard copies for a shard need to be requested
+     * @return the next shard copy
+     */
+    public SearchShardTarget findNext(final SearchShardIterator shardIt) {
+        SearchShardTarget next = shardIt.nextOrNull();
+        while (next != null && isWeighedAway(next.getNodeId())) {
```

Comment:
Can we simply check if there are shard routing that are not a part of the original iterator(without appended shards with weights zero) which can be used to retry/execute requests. The reason is there could be a race between when the request was first executed and when the request got retried(typically during concurrent failover requests)

T0 -> Weights(2:1:0)
T1 -> Requests failed on zone1 and then zone2
T2 -> Weights(0:1:2)
T3 -> Retry on third zone will be computed but this check(`isWeighedAway`) will fail this and prevent this request from executing successfully

Reply:
I realise that this check is put in place to avoid the shuffler from picking up zones that are weighed away, but can we do something to avoid independent check between when we add it to the iterator to when we actually consume it which is here

Reply:
yeah, this check is for weigh away zones, to avoid iterator to pick shards from weigh away az nodes in cases we don't want to fail open.




> but can we do something to avoid independent check between when we add it to the iterator to when we actually consume it which is here

I think most of the cases will have only one check but in worst case scenario where there is a need to fail open like cases we may have to iterate through all the shards and do `isWeighedAway` check. This won't be a very costly operation since size of the list is fixed.
Even if we precompute, weights can change in between which will require one more check.